OpenUtau.Plugin.Builtin/AdvancedPhonemizer.cs

```diff
@@ -0,0 +1,520 @@
+                        position = position,
+                        vowelTone = notes[noteI].tone
+                    };
+                    ccs = new List<string>();
+                    lastVowelI = lastSymbolI;
+                    syllableI++;
+                }
+            }
+
+            position += notes.Skip(vowelIds.Count).Sum(n => n.duration);
+
+            // making the ending
+            var lastNote = notes.Last();
+            word.ending = new Ending() {
+                prevV = word.syllables[syllableI - 1].v,
+                cc = symbols.Skip(lastVowelI + 1).ToArray(),
+                tone = lastNote.tone,
+                duration = lastNote.duration,
+                position = position + lastNote.duration
+            };
+
+            return word;
+        }
+
+        /// <summary>
+        /// Override this method, if you want to implement some machine converting from a word to phonemes
+        /// </summary>
+        /// <param name="word"></param>
+        /// <returns></returns>
+        protected virtual string[] HandleWordNotFound(string word) {
+            error = "word not found";
+            return null;
+        }
+
+        /// <summary>
+        /// Does this note extend the previous syllable?
+        /// </summary>
+        /// <param name="note"></param>
+        /// <returns></returns>
+        protected bool IsSyllableExtensionNote(Note note) {
+            return note.lyric.StartsWith("...~") || note.lyric.StartsWith("...*");
+        }
+
+        /// <summary>
+        /// May be used if you have different logic for short and long notes
+        /// </summary>
+        /// <param name="syllable"></param>
+        /// <returns></returns>
+        protected bool IsShort(Syllable syllable) {
+            return syllable.duration != -1 && TickToMs(syllable.duration) < GetTransitionBasicLength() * 2;
+        }
+        protected bool IsShort(Ending ending) {
+            return TickToMs(ending.duration) < GetTransitionBasicLength() * 2;
+        }
+
+        /// <summary>
+        /// Used to extract phonemes from CMU Dict word. Override if you need some extra logic
+        /// </summary>
+        /// <param name="phonemesString"></param>
+        /// <returns></returns>
+        protected virtual string[] GetDictionaryWordPhonemes(string phonemesString) {
+            return phonemesString.Split(' ');
+        }
+
+        /// <summary>
+        /// use to validate alias
+        /// </summary>
+        /// <param name="alias"></param>
+        /// <returns></returns>
+        protected virtual string ValidateAlias(string alias) {
+            return alias;
+        }
+
+        /// <summary>
+        /// multiply if some transitions are supposed to be longer or shorter
+        /// </summary>
+        /// <returns></returns>
+        protected virtual double GetTransitionBasicLength(string alias = "") {
+            return this.bpm < 190 ? 105 : this.bpm < 250 ? 90 : this.bpm < 300 ? 75 : 60;
```

Comment:
Rechecked Ms and Tick; added Ms or Tick postfixes to have this clear; also added a pre-made function to get the length by oto preutterance.

IsShort is a helper, it's not needed for base class itself, but it's used in final reclists, since there are some differences in use. I created helpers region and moved IsShort and HasOto there.